README.md
# TransIP Collection

[![CI](https://github.com/roaldnefs/ansible-collection-transip/workflows/CI/badge.svg?event=push)](https://github.com/roaldnefs/ansible-collection-transip/actions)

This collection contains modules and plugins to assist in automating [TransIP][transip] infrastructure and API interactions with Ansible

## Tested with Ansible

Tested with the current Ansible 2.9 and 2.10 release and the current development version of Ansible. Ansible versions before 2.9.10 are not supported.

## Included content

- transip_sshkey – Manage TransIP SSH keys
- transip_vps – Create and delete a TransIP VPS

## Installation and Usage

### Installing the Collection from Ansible Galaxy

Before using the TransIP collection, you need to install it with the Ansible Galaxy CLI:

    ansible-galaxy collection install roaldnefs.transip

You can also include it in a `requirements.yml` file and install it via `ansible-galaxy collection install -r requirements.yaml`, using the format:

```yaml
---
collections:
  - name: roaldnefs.transip
    version: 0.2.0
```

### Using modules from the TransIP Collection in your playbooks

It's preferable to use content in this collection using their Fully Qualified Collection Namespace (FQCN), for example: `roaldnefs.transip.transip_vps`:

```yaml
---
- hosts: localhost
  gather_facts: false
  connection: local

  task:
    - name: Create a new VPS
      roaldnefs.transip.transip_vps:
        state: present
        description: "example vps description"
        unique_description: yes
        product_name: vps-bladevps-x1
        operating_system: ubuntu-18.04
        access_token: REDACTED
      register: result
```

## Testing and Development

If you want to develop new content for this collection or improve what's already here, the easiest way to work on the collection is to clone it into one of the configured [`COLLECTIONS_PATHS`][ansible-collections-paths], ans work on it there.

## Testing with `ansible-test`

The `tests` directory contains configuration for running sanity and integeration tests using [`ansible-test`][ansible-test].

You can run the collection's test suites with the commands:

    ansible-test sanity --docker -v --color
    ansible-test integration --docker -v --color

Note: To run integration tests, you must add an [`integration_config.yml`][ansible-integration-config] file with a valid TransIP API key (using variable `transip_api_key`).

## Release notes

See the [changelog][changelog].

## Release process

Releases are automatically built and pushed to Ansible Galaxy for any new tag. Before tagging a release, make sure to do the following:

1. Update `galaxy.yml` and this README's `requirements.yml` example with the new `version` for the collection.
1. Update the CHANGELOG:
    1. Make sure you you have [`antsibull-changelog`][antsibull-changelog] installed.
    1. Make sure there are fragments for all known changes in `changelogs/fragments`.
    1. Run `antsibull-changelog release`.
1. Commit the changes and create a PR with the changes. Wait for tests to pass, then merge it once they have.
1. Tag the version in Git and push to GitHub.

After the version is published, verify it exists on the [TransIP Collection Galaxy page][ansible-galaxy-transip].

## More information

- [Ansible Collection overview](https://github.com/ansible-collections/overview)
- [Ansible User guide](https://docs.ansible.com/ansible/latest/user_guide/index.html)
- [Ansible Developer guide](https://docs.ansible.com/ansible/latest/dev_guide/index.html)
- [Ansible Community code of conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)

## Licensing

GNU General Public License v3.0 or later.

See [COPYING](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.

[transip]: https://www.transip.eu/
[changelog]: https://github.com/roaldnefs/transip-ansible-collection/blob/main/CHANGELOG.rst
[ansible-collections-paths]: https://docs.ansible.com/ansible/latest/reference_appendices/config.html#collections-paths
[ansible-test]: https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html
[ansible-integration-config]: https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html#integration-config-yml
[antsibull-changelog]: https://pypi.org/project/antsibull-changelog/
[ansible-galaxy-transip]: https://galaxy.ansible.com/roaldnefs/transip

CHANGELOG.rst
========================================
TransIP Ansible Collection Release Notes
========================================

.. contents:: Topics


v0.2.0
======

Major Changes
-------------

- transip_sshkey - add module to manage the SSH Keys in your TransIP account, these can be used to embed into your VPS at the time of its creation (https://github.com/roaldnefs/ansible-collection-transip/issues/2).

Bugfixes
--------

- transip_vps - fix retrieval of the API error message when creating a new VPS fails (https://github.com/roaldnefs/ansible-collection-transip/pull/4).

v0.1.1
======

Release Summary
---------------

Fix documentation and Ansible Galaxy configuration of the TransIP Collection.

v0.1.0
======

Release Summary
---------------

Initial release of the TransIP Collection.


plugins/module_utils/transip.py
# -*- coding: utf-8 -*-

# Copyright: (c) 2020, Roald Nefs <info@roaldnefs.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

import json
from ansible.module_utils.urls import fetch_url
from ansible.module_utils._text import to_text


class Response(object):

    def __init__(self, resp, info):
        self.body = None
        if resp:
            self.body = resp.read()
        self.info = info

    @property
    def json(self):
        if not self.body:
            return None
        try:
            return json.loads(to_text(self.body))
        except ValueError:
            return None

    @property
    def status_code(self):
        return self.info["status"]


class TransIPHelper(object):

    def __init__(self, module):
        self.module = module
        self.baseurl = "https://api.transip.nl/v6"
        self.timeout = module.params.get("timeout")
        self.access_token = module.params.get("access_token")
        self.headers = {"Authorization": "Bearer {0}".format(self.access_token),
                        "Content-type": "application/json"}

        # Call the simple test resource on the TransIP API to make sure
        # everything is working
        response = self.get("api-test")
        if response.status_code == 401:
            self.module.fail_json(msg="Failed to login using API token, please verify validity of the API token.")

    @staticmethod
    def transip_argument_spec():
        return dict(
            access_token=dict(
                no_log=True,
                required=False,
                aliases=["api_token"]
            ),
            timeout=dict(type='int', default=30),
        )

    def send(self, method, path, data=None):
        url = "{0}/{1}".format(self.baseurl, path)
        data = self.module.jsonify(data)

        resp, info = fetch_url(
            self.module,
            url,
            data=data,
            headers=self.headers,
            method=method,
            timeout=self.timeout
        )

        return Response(resp, info)

    def get(self, path, data=None):
        return self.send("GET", path, data)

    def post(self, path, data=None):
        return self.send("POST", path, data)

    def put(self, path, data=None):
        return self.send("PUT", path, data)

    def patch(self, path, data=None):
        return self.send("PATCH", path, data)

    def delete(self, path, data=None):
        return self.send("DELETE", path, data)


plugins/modules/transip_domain.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2020, Roald Nefs <info@roaldnefs.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import absolute_import, division, print_function
__metaclass__ = type

DOCUMENTATION = r'''
---
module: transip_domain
short_description:
description:
author: "Roald Nefs (@roaldnefs)"
options:
extends_documentation_fragment:
- roaldnefs.transip.transip.documentation
'''

import traceback
from ansible.module_utils.basic import AnsibleModule
from ansible_collections.roaldnefs.transip.plugins.module_utils.transip import TransIPHelper
from ansible.module_utils._text import to_native


def core(module):
    pass


def main():
    argument_spec = TransIPHelper.transip_argument_spec()
    argument_spec.update(
        state=dict(choices=["present"], default="present")
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
    )

    try:
        core(module)
    except Exception as exc:
        module.fail_json(msg=to_native(exc), exception=traceback.format_exc())

if __name__ == "__main__":
    main()


plugins/modules/transip_sshkey.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2020, Roald Nefs <info@roaldnefs.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import absolute_import, division, print_function
__metaclass__ = type


DOCUMENTATION = r'''
---
module: transip_sshkey
short_description: Manage TransIP SSH keys
description:
- Manage TransIP SSH keys.
author: "Roald Nefs (@roaldnefs)"
options:
  state:
    description:
    - Indicate desired state of the target.
    default: present
    choices: ['present', 'absent']
    type: str
  description:
    description:
    - The SSH key description.
    type: str
  fingerprint:
    description:
    - The MD5 fingerprint of the SSH key.
    type: str
  ssh_pub_key:
    description:
    - The SSH key.
    type: str
extends_documentation_fragment:
- roaldnefs.transip.transip.documentation
'''

EXAMPLES = r'''
- name: Create a new SSH key
  roaldnefs.transip.transip_sshkey:
    state: present
    ssh_pub_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDf2pxWX/yhUBDyk2LPhvRtI0LnVO8PyR5Zt6AHrnhtLGqK+8YG9EMlWbCCWrASR+Q1hFQG example
    description: example
  register: result

- debug:
    msg: "Added SSH-key with fingerprint {{ result.data.sshKey.fingerprint }}"

- name: Delete a SSH key
  roaldnefs.transip.transip_sshkey:
    state: absent
    ssh_pub_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDf2pxWX/yhUBDyk2LPhvRtI0LnVO8PyR5Zt6AHrnhtLGqK+8YG9EMlWbCCWrASR+Q1hFQG example
'''

RETURN = r'''
data:
  description: a TransIP SSH key
  returned: changed
  type: dict
  sample: {
    "sshKey": {
        "creationDate": "2020-12-27 15:24:59",
        "description": "example",
        "fingerprint": "79:07:8c:7c:6c:00:b1:e9:44:4a:bf:e4:1a:fa:88:0d",
        "id": 10163,
        "key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDf2pxWX/yhUBDyk2LPhvRtI0LnVO8PyR5Zt6AHrnhtLGqK+8YG9EMlWbCCWrASR+Q1hFQG example"
    }
  }
'''


import traceback
from ansible.module_utils.basic import AnsibleModule
from ansible_collections.roaldnefs.transip.plugins.module_utils.transip import TransIPHelper


class TransIPSSHKeyException(Exception):
    pass


class TransIPSSHKey(object):
    def __init__(self, module):
        self.module = module
        self.rest = TransIPHelper(module)

    def list(self):
        """Returns a list of all SSH-keys."""
        path = "ssh-keys"
        response = self.rest.get(path)

        if response.status_code == 200:
            return response.json.get("sshKeys", [])

        return []

    def get(self):
        """Get information of a specific SSH-key."""
        fingerprint = self.module.params.get("fingerprint")
        ssh_pub_key = self.module.params.get("ssh_pub_key")

        for ssh_key in self.list():
            if fingerprint:
                if fingerprint == ssh_key["fingerprint"]:
                    return {"sshKey": ssh_key}
            else:
                if ssh_pub_key == ssh_key["key"]:
                    return {"sshKey": ssh_key}
        return None

    def create(self):
        """Create a new SSH-key if it doesn't exists yet."""
        json_data = self.get()
        if json_data:
            self.module.exit_json(changed=False)
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        path = "ssh-keys"

        # Set the required request attributes
        data = {
            "sshKey": self.module.params.get("ssh_pub_key"),
            "description": self.module.params.get("description")
        }

        response = self.rest.post(path, data=data)

        if response.status_code == 201:
            # Retrieve data about the newly created SSH-key, as the TransIP API
            # doesn't provide any information on the creation of the SSH-key
            json_data = self.get()
            if json_data and "sshKey" in json_data:
                ssh_key_data = json_data
            else:
                # When using the demo access token, the API doesn't actually
                # create the SSH-key so the json_data might be empty
                ssh_key_data = {"sshKey": {}}
            self.module.exit_json(changed=True, data=ssh_key_data)
        else:
            json_data = response.json
            error_msg = "Failed to create SSH-key"
            if json_data and "error" in json_data:
                error_msg = json_data["error"]
            self.module.fail_json(change=False, msg=error_msg)

    def delete(self):
        """Delete an existing SSH-key."""
        json_data = self.get()
        if not json_data:
            self.module.exit_json(changed=False, msg="SSH-key not found")

        if self.module.check_mode:
            self.module.exit_json(changed=True)

        ssh_key_id = json_data["sshKey"]["id"]
        path = "sshKeys/{0}".format(ssh_key_id)

        response = self.rest.delete(path)

        if response.status_code == 204:
            self.module.exit_json(changed=True, msg="SSH-key deleted")
        else:
            self.module.fail_json(changed=False, msg="Failed to delete SSH-key")


def handle_request(module):
    sshkey = TransIPSSHKey(module)
    state = module.params["state"]
    if state == "present":
        sshkey.create()
    elif state == "absent":
        sshkey.delete()


def main():
    argument_spec = TransIPHelper.transip_argument_spec()
    argument_spec.update(
        state=dict(choices=["present", "absent"], default="present"),
        ssh_pub_key=dict(type="str"),
        description=dict(type="str"),
        fingerprint=dict(type="str"),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
        required_one_of=(
            ("ssh_pub_key", "fingerprint"),
        ),
        required_if=(
            ("state", "present", ["ssh_pub_key", "description"]),
        ),
    )

    try:
        handle_request(module)
    except TransIPSSHKeyException as exc:
        module.fail_json(msg=str(exc), exception=traceback.format_exc())
    except KeyError as exc:
        module.fail_json(msg='Unable to load {0}'.format(str(exc)), exception=traceback.format_exc())


if __name__ == "__main__":
    main()


plugins/modules/transip_vps.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright: (c) 2020, Roald Nefs <info@roaldnefs.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import absolute_import, division, print_function
__metaclass__ = type


DOCUMENTATION = r'''
---
module: transip_vps
short_description: Create and delete a TransIP VPS
description:
- Create and delete a VPS in TransIP.
author: "Roald Nefs (@roaldnefs)"
options:
  state:
    description:
    - Indicate desired state of the target.
    default: present
    choices: ['present', 'absent']
    type: str
  name:
    description:
    - The unique name of the VPS as provided by TransIP.
    type: str
  description:
    description:
    - The name of the VPS that can be set by the customer.
    type: str
  unique_description:
    description:
    - Require an unique description of the VPS. By default TransIP allows multiple hosts with the
      same description. Setting this to "yes" allows only unique descriptions to be used. Useful for
      idempotence, as the unique name of the VPS will be generated by TransIP.
    default: False
    type: bool
  product_name:
    description:
    - Name of the product, e.g. vps-bladevps-x1.
    type: str
  operating_system:
    description:
    - The name of the operating system to install, e.g. ubuntu-18.04.
    type: str
  end_time:
    description:
    - Indicate when the VPS will be terminated.
    - On 'end', the VPS will be terminated on the end date of the agreement.
    - On 'immediately', the VPS will be terminated immediately.
    default: end
    choices: ['end', 'immediately']
    type: str
extends_documentation_fragment:
- roaldnefs.transip.transip.documentation
'''

EXAMPLES = r'''
- name: Create a new VPS
  roaldnefs.transip.transip_vps:
    state: present
    description: "example vps description"
    unique_description: yes
    product_name: vps-bladevps-x1
    operating_system: ubuntu-18.04
    access_token: REDACTED
  register: result

- debug:
    msg: "Created new VPS with name {{ result.data.vps.name }}."

- name: Delete a VPS
  roaldnefs.transip.transip_vps:
    state: absent
    name: transipdemo-vps
    end_time: immediately
    access_token: REDACTED
'''

RETURN = r'''
data:
  description: a TransIP VPS
  returned: changed
  type: dict
  sample: {
    "vps": {
        "availabilityZone": "ams0",
        "cpus": 3,
        "currentSnapshots": 1,
        "description": "",
        "diskSize": 52428800,
        "ipAddress": "141.138.136.129",
        "isBlocked": false,
        "isCustomerLocked": false,
        "isLocked": false,
        "macAddress": "52:54:00:19:a7:20",
        "maxSnapshots": 1,
        "memorySize": 1048576,
        "name": "transipdemo-vps",
        "operatingSystem": "FreeBSD 10.0-RELEASE",
        "productName": "vps-bladevps-x1",
        "status": "running",
        "tags": [
            "customTag",
            "anotherTag"
        ]
    }
  }
'''

import traceback
from ansible.module_utils.basic import AnsibleModule
from ansible_collections.roaldnefs.transip.plugins.module_utils.transip import TransIPHelper


class TransIPVPSException(Exception):
    pass


class TransIPVPS(object):

    def __init__(self, module):
        self.module = module
        self.rest = TransIPHelper(module)

    def list(self):
        """Returns a list of all VPSs."""
        path = "vps"
        response = self.rest.get(path)

        if response.status_code == 200:
            return response.json.get("vpss", [])

        return []

    def get(self):
        """Get information of a specific VPS."""
        name = self.module.params.get("name")
        description = self.module.params.get("description")
        unique_description = self.module.params.get("unique_description")

        json_data = self.get_by_name(name)
        if not json_data and unique_description:
            json_data = self.get_by_description(description)
        return json_data

    def get_by_name(self, name):
        """Get information of a specific VPS by name."""
        path = "vps/{0}".format(name)
        response = self.rest.get(path)

        if response.status_code == 200:
            return response.json

        return None

    def get_by_description(self, description):
        """Get information of a specific VPS by description."""
        for vps in self.list():
            if vps["description"] == description:
                return {"vps": vps}
        return None

    def create(self):
        """Order a new VPS if it doesn't already exists."""
        json_data = self.get()
        if json_data:
            self.module.exit_json(changed=False)
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        path = "vps"

        # Set the required request attributes
        data = {
            "productName": self.module.params["product_name"],
            "operatingSystem": self.module.params["operating_system"],
        }

        # Add the optional request attributes
        description = self.module.params.get("description")
        if description:
            data["description"] = description

        response = self.rest.post(path, data=data)

        if response.status_code == 201:
            # Retrieve data about the newly created VPS, as the TransIP API
            # doesn't provide any information on the creation of the VPS
            json_data = self.get()
            # When using the demo access token, the API doesn't actually create
            # the VPS so the json_data might be empty
            vps_data = {"vps": {}}
            if json_data and "vps" in json_data:
                vps_data["vps"] = json_data["vps"]
            self.module.exit_json(changed=True, data=vps_data)
        else:
            json_data = response.json
            error_msg = "Failed to order VPS"
            if json_data and "error" in json_data:
                error_msg = json_data["error"]
            self.module.fail_json(changed=False, msg=error_msg)

    def cancel(self):
        """Cancel a VPS using the specified end time attribute, on 'end' will
        terminate the VPS on the end date of the agreement while 'immediately'
        will terminate the VPS immediately.
        """
        name = self.module.params.get("name")
        end_time = self.module.params.get("end_time")

        json_data = self.get_by_name(name)
        if json_data:
            if self.module.check_mode:
                self.module.exit_json(change=True)
            path = "vps/{0}".format(name)
            data = {"endTime": end_time}
            response = self.rest.delete(path, data=data)

            if response.status_code == 204:
                self.module.exit_json(changed=True, msg="VPS canceled")
            self.module.fail_json(changed=False, msg="Failed to cancel VPS")
        else:
            self.module.fail_json(changed=False, msg="VPS not found")


def handle_request(module):
    vps = TransIPVPS(module)
    state = module.params["state"]
    if state == "present":
        vps.create()
    elif state == "absent":
        vps.cancel()


def main():
    argument_spec = TransIPHelper.transip_argument_spec()
    argument_spec.update(
        state=dict(choices=["present", "absent"], default="present"),
        name=dict(type="str"),
        product_name=dict(type="str"),
        operating_system=dict(type="str"),
        end_time=dict(choices=["end", "immediately"], default="end"),
        description=dict(type="str"),
        unique_description=dict(type="bool", default=False),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        required_if=([
            ("state", "present", ["product_name", "operating_system"]),
            ("state", "absent", ["end_time"]),
            ("unique_description", True, ["description"]),
        ]),
        supports_check_mode=True,
    )

    try:
        handle_request(module)
    except TransIPVPSException as exc:
        module.fail_json(msg=str(exc), exception=traceback.format_exc())
    except KeyError as exc:
        module.fail_json(msg='Unable to load {0}'.format(str(exc)), exception=traceback.format_exc())


if __name__ == "__main__":
    main()
